master/internal/project/postgres_project.go

```diff
@@ -50,3 +136,285 @@ func ProjectIDByName(ctx context.Context, workspaceID int, projectName string) (
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			return db.ErrNotFound
+		} else if err != nil {
+			return err
+		}
+
+		project, err = GetProjectByID(ctx, projectID)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	// GetProjectByID handles the case where the project is not found.
+	if err != nil {
+		return nil, err
+	}
+	return project, nil
+}
+
+// GenerateProjectKey generates a unique project key for a project based on its name.
+func generateProjectKey(ctx context.Context, tx bun.Tx, projectName string) (string, error) {
+	var key string
+	found := true
+	for i := 0; i < MaxRetries && found; i++ {
+		prefixLength := min(len(projectName), MaxProjectKeyPrefixLength)
+		prefix := projectName[:prefixLength]
+		suffix := random.String(MaxProjectKeyLength - prefixLength)
+		key = strings.ToUpper(prefix + suffix)
+		err := tx.NewSelect().Model(&model.Project{}).Where("key = ?", key).For("UPDATE").Scan(ctx)
+		found = err == nil
+	}
+	if found {
+		return "", fmt.Errorf("could not generate a unique project key")
+	}
+	return key, nil
+}
+
+// InsertProject inserts a new project into the database.
+func InsertProject(
+	ctx context.Context,
+	p *model.Project,
+	requestedKey *string,
+) (err error) {
+RetryLoop:
+	for i := 0; i < MaxRetries; i++ {
+		err = db.Bun().RunInTx(ctx, &sql.TxOptions{Isolation: sql.LevelRepeatableRead},
+			func(ctx context.Context, tx bun.Tx) error {
+				var err error
+				if requestedKey == nil {
+					p.Key, err = generateProjectKey(ctx, tx, p.Name)
+					if err != nil {
+						return err
+					}
+				} else {
+					p.Key = *requestedKey
+				}
+				_, err = tx.NewInsert().Model(p).Exec(ctx)
+
+				if err != nil && strings.Contains(err.Error(), db.CodeUniqueViolation) {
+					switch errString := err.Error(); {
+					case strings.Contains(errString, "projects_key_key"):
+						return errors.Wrapf(db.ErrDuplicateRecord, "project key %s is already in use", p.Key)
+					case strings.Contains(errString, "projects_name_workspace_id_key"):
+						return errors.Wrapf(db.ErrDuplicateRecord, "project name %s is already in use", p.Name)
+					default:
+						return err
+					}
+				} else if err != nil {
+					return err
+				}
+				return nil
+			},
+		)
+
+		switch {
+		case err == nil:
+			break RetryLoop
+		case requestedKey == nil && strings.Contains(err.Error(), fmt.Sprintf("project key %s is already in use", p.Key)):
+			log.Debugf("retrying project (%s) insertion due to generated key conflict (%s)", p.Name, p.Key)
+			continue // retry
+		default:
+			break RetryLoop
+		}
+	}
+	return errors.Wrapf(err, "error inserting project %s into database", p.Name)
+}
+
+// UpdateProjectKey updates the key of a project.
+func UpdateProjectKey(ctx context.Context, projectID int, key string) error {
+	if len(key) > MaxProjectKeyLength {
+		return fmt.Errorf("project key must be at most %d characters", MaxProjectKeyLength)
+	}
+	if projectID == 0 {
+		return fmt.Errorf("invalid project ID")
+	}
```

Comment:
Can we call `ValidateProjectKey` here? Also, where is this used?

Reply:
Good call out! This approach had been swapped out to follow suit with PatchProject, so this function can be dropped.